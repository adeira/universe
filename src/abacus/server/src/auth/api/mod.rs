use crate::auth::dal::users::list_all_users;
use crate::auth::rbac;
use crate::auth::rbac::Actions::Users;
use crate::auth::rbac::UsersActions::GetAllUsers;
use crate::auth::users::{AnyUser, User};
use crate::graphql::AbacusGraphQLResult;
use crate::graphql_context::Context;
use juniper::FieldResult;

pub(crate) struct AuthQuery;
pub(crate) struct AuthMutation;

#[juniper::graphql_object(context = Context)]
impl AuthQuery {
    /// Returns information about the current user (can be authenticated or anonymous).
    async fn whoami(context: &Context) -> crate::auth::api::WhoamiPayload {
        crate::auth::api::whoami(context).await
    }

    async fn list_users(context: &Context) -> AbacusGraphQLResult<Vec<AnyUser>> {
        Ok(crate::auth::api::list_users(context).await?)
    }
}

#[juniper::graphql_object(context = Context)]
impl AuthMutation {
    /// This function accepts Google ID token (after receiving it from Google Sign-In in a webapp)
    /// and returns authorization payload. There is no concept of sign-in and sign-up: every
    /// whitelisted user with a valid JWT ID token will be authorized. Invalid tokens and users
    /// that are not whitelisted will be rejected.
    ///
    /// Repeated calls will result in a new session token (old sessions are not removed to support
    /// multiple logins on multiple devices). Original session token is returned back only once
    /// and cannot be retrieved later (it's irreversibly hashed in the database).
    async fn authorize_webapp(
        google_id_token: String,
        context: &Context,
    ) -> FieldResult<crate::auth::api::AuthorizeWebappPayload> {
        crate::auth::api::authorize_webapp(&google_id_token, context).await
    }

    /// The purpose of this `deauthorize` mutation is to remove the active sessions and effectively
    /// make the mobile application/webapp unsigned. Applications should remove the session token
    /// once de-authorized.
    ///
    /// Repeated calls will result in failure since it's not possible to deauthorize twice.
    async fn deauthorize(
        session_token: String, // TODO: this could be removed (?) - we can use the user from context
        context: &Context,
    ) -> crate::auth::api::DeauthorizePayload {
        crate::auth::api::deauthorize(&session_token, context).await
    }
}

#[derive(juniper::GraphQLObject)]
pub(crate) struct WhoamiPayload {
    id: Option<juniper::ID>,

    /// Human readable type should be used only for testing purposes. The format is not guaranteed
    /// and can change in the future completely.
    human_readable_type: Option<String>,

    /// Debug assertions indicates that the Rust server runs in a development mode (compiled without
    /// optimizations). FOR DEVELOPMENT ONLY!
    is_debug_assertions_enabled: bool,
}

pub(crate) async fn whoami(context: &Context) -> WhoamiPayload {
    match &context.user {
        User::SignedUser(user) => WhoamiPayload {
            id: Some(juniper::ID::from(user.id())),
            human_readable_type: Some(String::from("signed user")),
            is_debug_assertions_enabled: cfg!(debug_assertions),
        },
        User::AnonymousUser(user) => WhoamiPayload {
            id: Some(juniper::ID::from(user.id())),
            human_readable_type: Some(String::from("anonymous user")),
            is_debug_assertions_enabled: cfg!(debug_assertions),
        },
    }
}

pub(crate) async fn list_users(context: &Context) -> anyhow::Result<Vec<AnyUser>> {
    match rbac::verify_permissions(&context.user, &Users(GetAllUsers)).await {
        // only admin can list all the users
        Ok(_) => match list_all_users(&context.pool).await {
            Ok(list) => Ok(list),
            Err(e) => Err(e),
        },
        Err(_) => anyhow::bail!("not enough permissions to list all the users"),
    }
}

///// Mutations:

#[derive(juniper::GraphQLObject)]
pub(crate) struct AuthorizeWebappPayload {
    success: bool,

    /// Failure message is available only when success=false.
    failure_message: Option<String>,

    /// Session token should be send with every GraphQL request which requires auth.
    /// Returns `None` if the request was not successful.
    session_token: Option<String>,
}

#[derive(juniper::GraphQLObject)]
pub(crate) struct DeauthorizePayload {
    success: bool,
}

/// Webapp is currently using very simplified and minimalistic flow: it basically allows only
/// specific Google subjects to be authorized (admins). The thinking is that webapp login is needed
/// only for the backoffice and the access should be limited to admins. This should be easily
/// changeable in the future.
pub(crate) async fn authorize_webapp(
    google_id_token: &str,
    context: &Context,
) -> FieldResult<AuthorizeWebappPayload> {
    let connection_pool = context.pool.to_owned();
    let session_token = crate::auth::authorize(&connection_pool, google_id_token).await;
    match session_token {
        Ok(session_token) => Ok(AuthorizeWebappPayload {
            success: true,
            failure_message: None,
            session_token: Some(session_token),
        }),
        Err(e) => {
            tracing::error!("{}", e);
            Ok(AuthorizeWebappPayload {
                success: false,
                failure_message: Some(format!("{}", e)),
                session_token: None,
            })
        }
    }
}

pub(crate) async fn deauthorize(
    session_token: &str, // TODO: this could be removed (?) - we can use the user from context
    context: &Context,
) -> DeauthorizePayload {
    let connection_pool = context.pool.to_owned();
    match crate::auth::deauthorize(&connection_pool, session_token).await {
        Ok(_) => DeauthorizePayload { success: true },
        Err(_) => DeauthorizePayload { success: false },
    }
}
