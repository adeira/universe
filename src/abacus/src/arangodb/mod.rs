use crate::arangodb::pool::ConnectionManager;
use arangors::{ArangoError, ClientError, Connection};
#[cfg(test)]
use deadpool::managed::Object;
use deadpool::managed::Pool;

pub mod errors;
mod pool;

#[derive(Clone)]
pub struct ConnectionPool {
    pub pool: Pool<ConnectionManager>,
    db_name: String,
}

type Database = arangors::Database<uclient::reqwest::ReqwestClient>;

impl ConnectionPool {
    pub async fn db(&self) -> Database {
        let status = &self.pool.status();
        tracing::trace!(
            "Connection pool status: max_size={}, size={}, available={}",
            // The maximum size of the pool
            status.max_size,
            // The current size of the pool
            status.size,
            // The number of available objects in the pool. If there are no objects in the pool
            // this number can become negative and stores the number of futures waiting for an object.
            status.available
        );

        // We use `unwrap` (`expect`) here because not being able to access the database is an
        // unrecoverable situation and we want to panic. Not sure what to do otherwise (?).
        let connection = &self
            .pool
            .get()
            .await
            .expect("could not get database connection from the pool");

        get_or_create_db(&connection, &self.db_name).await
    }

    #[cfg(test)]
    pub async fn connection(&self) -> Object<ConnectionManager> {
        self.pool
            .get()
            .await
            .expect("could not get database connection from the pool")
    }
}

async fn get_or_create_db(connection: &Connection, db_name: &str) -> Database {
    match connection.db(db_name).await {
        Ok(database) => database,
        Err(ClientError::Arango(ArangoError { .. })) => {
            // TODO: check 1228 - ERROR_ARANGO_DATABASE_NOT_FOUND code
            // https://www.arangodb.com/docs/stable/appendix-error-codes.html
            tracing::info!("ðŸ¥¶ Cold start - creating a new database ðŸ¥¶");
            connection
                .create_database(db_name)
                .await
                .unwrap_or_else(|_| panic!("could not create '{}' database", db_name))
        }
        Err(error) => {
            panic!("could not access '{}' database: {:?}", db_name, error)
        }
    }
}

/// Database name `None` indicates that the test doesn't actually need a database and should fail
/// if the application tries to access it anyway.
pub fn get_database_connection_pool(
    arangodb_url: &str,
    arangodb_database: &str,
    arangodb_username: &str,
    arangodb_password: &str,
) -> ConnectionPool {
    // Maximum number of connections ever created.
    // TODO: what should be the actual maximum size?
    // https://github.com/arangodb/arangodb/blob/35c278cdf3b7985f8ed2042dfef8d22c2dd2ed07/arangod/Network/ConnectionPool.h#L65
    let max_pool_size = num_cpus::get_physical() * 4;

    let connection_pool = deadpool::managed::Pool::new(
        pool::ConnectionManager {
            db_host: arangodb_url.to_string(),
            db_name: arangodb_database.to_string(),
            username: arangodb_username.to_string(),
            password: arangodb_password.to_string(),
        },
        max_pool_size,
    );

    tracing::trace!(
        "Creating (empty) database connection pool for: '{}' ({}) ðŸ”¥",
        arangodb_database,
        arangodb_url
    );
    ConnectionPool {
        pool: connection_pool,
        db_name: arangodb_database.to_string(),
    }
}

#[cfg(test)]
pub fn get_database_connection_pool_mock() -> ConnectionPool {
    get_database_connection_pool(
        "mock_arangodb_url",
        "mock_arangodb_database",
        "mock_arangodb_username",
        "mock_arangodb_password",
    )
}

/// Creates an empty database for test purposes. Each test should define custom `db_name` to
/// avoid conflicts with other tests. The following steps are performed:
///
/// 1) the DB is deleted (if it already exists for some reason)
/// 2) database migrations are applied
///
/// TODO: how to properly construct the `db_name` to keep it sane?
#[cfg(test)]
pub async fn prepare_empty_test_database(db_name: &str) -> ConnectionPool {
    cleanup_test_database(db_name).await;
    let pool = get_database_connection_pool(
        // TODO: make it more DX/test friendly (?)
        // TODO: dev k8s cluster
        "http://arangodb-single-server.default.svc.cluster.local:8529",
        db_name,
        "",
        "",
    );
    crate::migrations::migrate(&pool).await;
    pool
}

/// This function cleanup the test database by removing it completely.
#[cfg(test)]
pub async fn cleanup_test_database(db_name: &str) -> ConnectionPool {
    let pool = get_database_connection_pool(
        // TODO: make it more DX/test friendly (?)
        // TODO: dev k8s cluster
        "http://arangodb-single-server.default.svc.cluster.local:8529",
        db_name,
        "",
        "",
    );
    let connection = pool.connection().await;
    if connection.db(&db_name).await.is_ok() {
        // database already exists, let's delete it
        connection
            .drop_database(db_name) // Should we throw if it already exists?
            .await
            .unwrap_or_else(|_| panic!("Could not delete the test database: {}", db_name));
    }
    pool
}
