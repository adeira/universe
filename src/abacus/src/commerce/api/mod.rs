pub use crate::commerce::model::errors::ModelError;
pub use crate::commerce::model::products::PriceSortDirection;
pub use crate::commerce::model::products::Product;
pub use crate::commerce::model::products::ProductMultilingualInput;
pub use crate::commerce::model::products::ProductMultilingualInputVisibility;
pub use crate::commerce::model::products::SupportedCurrency;
pub use crate::commerce::model::products::SupportedLocale;

use crate::graphql_context::Context;

#[derive(juniper::GraphQLObject)]
pub struct ProductError {
    message: String,
}

#[derive(juniper::GraphQLUnion)]
pub enum ProductOrError {
    Product(Product),
    ProductError(ProductError),
}

pub(crate) struct CommerceQuery;
pub(crate) struct CommerceMutation;

#[juniper::graphql_object(context = Context)]
impl CommerceQuery {
    /// Searches all published (publicly accessible) products. Everyone can do it without any
    /// special permission so it should be used on FE.
    async fn search_published_products(
        context: &Context,
        client_locale: SupportedLocale,
        price_sort_direction: PriceSortDirection,
        search_term: Option<String>,
    ) -> Option<Vec<Option<Product>>> {
        // TODO: return Result<â€¦, ModelError>
        search_published_products(
            &context,
            &client_locale,
            &price_sort_direction,
            &search_term,
            &ProductMultilingualInputVisibility::ESHOP,
        )
        .await
    }

    /// Searches all products (published and unpublished). Requires admin permissions so it should
    /// be used only in backoffice to administer the products.
    async fn search_all_products(
        context: &Context,
        client_locale: SupportedLocale,
        price_sort_direction: PriceSortDirection,
        search_term: Option<String>,
    ) -> Result<Vec<Option<Product>>, ModelError> {
        crate::commerce::model::products::search_all_products(
            &context,
            &client_locale,
            &price_sort_direction,
            &search_term,
        )
        .await
    }

    /// Returns one publicly available product by its key. Anyone can call this resolver.
    async fn get_published_product_by_key(
        context: &Context,
        client_locale: SupportedLocale,
        product_key: juniper::ID,
    ) -> Result<Product, ModelError> {
        crate::commerce::model::products::get_published_product_by_key(
            &context,
            &client_locale,
            &product_key,
        )
        .await
    }

    /// Only admins can call this function! It returns published OR unpublished product by its key.
    async fn get_unpublished_product_by_key(
        context: &Context,
        client_locale: SupportedLocale,
        product_key: juniper::ID,
    ) -> Result<Product, ModelError> {
        crate::commerce::model::products::get_unpublished_product_by_key(
            &context,
            &client_locale,
            &product_key,
        )
        .await
    }
}

#[juniper::graphql_object(context = Context)]
impl CommerceMutation {
    /// Creates a new product.
    ///
    /// Note on uploading product images: image names specified in the GraphQL input must correspond
    /// to the uploadables (multipart/form-data) and vice versa. Requests with invalid uploadables
    /// will be rejected.
    async fn product_create(
        context: &Context,
        product_multilingual_input: ProductMultilingualInput,
    ) -> ProductOrError {
        match crate::commerce::model::products::create_product(
            &context,
            &product_multilingual_input,
        )
        .await
        {
            Ok(product) => ProductOrError::Product(product),
            Err(e) => ProductOrError::ProductError(ProductError {
                // TODO: do not expose DB and RBAC errors directly
                message: e.to_string(),
            }),
        }
    }

    /// Updates already existing product with new values. It requires not only product KEY but also
    /// product REVISION to avoid lost update situations (when someone else tried to update the
    /// product and this update would overwrite the latest changes).
    ///
    /// Note on updating product images: already existing image names must be send to the server
    /// otherwise they will be deleted. You can optionally specify some extra (new) images to upload
    /// them via uploadables. This feature will eventually be used even for images re-ordering.
    async fn product_update(
        context: &Context,
        product_key: juniper::ID,
        product_revision: juniper::ID,
        product_multilingual_input: ProductMultilingualInput,
    ) -> ProductOrError {
        match crate::commerce::model::products::update_product(
            &context,
            &product_key,
            &product_revision,
            &product_multilingual_input,
        )
        .await
        {
            Ok(product) => ProductOrError::Product(product),
            Err(e) => ProductOrError::ProductError(ProductError {
                // TODO: do not expose DB and RBAC errors directly
                message: e.to_string(),
            }),
        }
    }

    /// Deletes product based on the product KEY.
    async fn product_delete(context: &Context, product_key: juniper::ID) -> ProductOrError {
        match crate::commerce::model::products::delete_product(&context, &product_key).await {
            Ok(product) => ProductOrError::Product(product),
            Err(e) => ProductOrError::ProductError(ProductError {
                // TODO: do not expose DB and RBAC errors directly
                message: e.to_string(),
            }),
        }
    }

    /// Publishes product based on the product KEY. Various validation requirements must be met
    /// before the product can be published. Published product is available outside of backoffice.
    async fn product_publish(context: &Context, product_key: juniper::ID) -> ProductOrError {
        match crate::commerce::model::products::publish_product(&context, &product_key).await {
            Ok(product) => ProductOrError::Product(product),
            Err(e) => ProductOrError::ProductError(ProductError {
                // TODO: do not expose DB and RBAC errors directly
                message: e.to_string(),
            }),
        }
    }

    /// Unpublishes product based on the product KEY. Unpublished products are available only inside
    /// the backoffice.
    async fn product_unpublish(context: &Context, product_key: juniper::ID) -> ProductOrError {
        match crate::commerce::model::products::unpublish_product(&context, &product_key).await {
            Ok(product) => ProductOrError::Product(product),
            Err(e) => ProductOrError::ProductError(ProductError {
                // TODO: do not expose DB and RBAC errors directly
                message: e.to_string(),
            }),
        }
    }
}

// This function is exposed to GraphQL commerce module as well as to POS module (hence not inlined).
pub(crate) async fn search_published_products(
    context: &Context,
    client_locale: &SupportedLocale,
    price_sort_direction: &PriceSortDirection,
    search_term: &Option<String>,
    visibility: &ProductMultilingualInputVisibility,
) -> Option<Vec<Option<Product>>> {
    match crate::commerce::model::products::search_published_products(
        &context,
        &client_locale,
        &price_sort_direction,
        &search_term,
        &visibility,
    )
    .await
    {
        Ok(products) => Some(products),
        Err(e) => {
            tracing::error!("{}", e);
            None
        }
    }
}

pub(crate) async fn get_products_by_keys(
    context: &Context,
    client_locale: &SupportedLocale,
    product_keys: &[String],
) -> Result<Vec<Product>, ModelError> {
    crate::commerce::model::products::get_published_products_by_keys(
        &context,
        &client_locale,
        &product_keys,
    )
    .await
}
