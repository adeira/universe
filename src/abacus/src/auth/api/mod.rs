use crate::arangodb::errors::ModelError;
use crate::auth::dal::users::list_all_users;
use crate::auth::rbac;
use crate::auth::rbac::Actions::Users;
use crate::auth::rbac::UsersActions::GetAllUsers;
use crate::auth::users::{AnyUser, User};
use crate::graphql_context::Context;
use juniper::FieldResult;

#[derive(juniper::GraphQLObject)]
pub(crate) struct WhoamiPayload {
    id: Option<juniper::ID>,

    /// Human readable type should be used only for testing purposes. The format is not guaranteed
    /// and can change in the future completely.
    human_readable_type: Option<String>,

    /// Debug assertions indicates that the Rust server runs in a development mode (compiled without
    /// optimizations). FOR DEVELOPMENT ONLY!
    is_debug_assertions_enabled: bool,
}

pub(crate) async fn whoami(context: &Context) -> WhoamiPayload {
    match &context.user {
        User::SignedUser(user) => WhoamiPayload {
            id: Some(juniper::ID::from(user.id())),
            human_readable_type: Some(String::from("signed user")),
            is_debug_assertions_enabled: cfg!(debug_assertions),
        },
        User::AnonymousUser(user) => WhoamiPayload {
            id: Some(juniper::ID::from(user.id())),
            human_readable_type: Some(String::from("anonymous user")),
            is_debug_assertions_enabled: cfg!(debug_assertions),
        },
    }
}

pub(crate) async fn list_users(context: &Context) -> Result<Vec<AnyUser>, ModelError> {
    match rbac::verify_permissions(&context.user, &Users(GetAllUsers)).await {
        // only admin can list all the users
        Ok(_) => match list_all_users(&context.pool).await {
            Ok(list) => Ok(list),
            Err(e) => Err(e),
        },
        Err(_) => Err(ModelError::PermissionsError(String::from(
            "not enough permissions to list all the users",
        ))),
    }
}

///// Mutations:

#[derive(juniper::GraphQLObject)]
pub(crate) struct AuthorizeMobilePayload {
    success: bool,

    /// Failure message is available only when success=false.
    failure_message: Option<String>,

    /// Session token should be send with every GraphQL request which requires auth.
    /// Returns `None` if the request was not successful.
    session_token: Option<String>,
}

#[derive(juniper::GraphQLObject)]
pub(crate) struct AuthorizeWebappPayload {
    success: bool,

    /// Failure message is available only when success=false.
    failure_message: Option<String>,

    /// Session token should be send with every GraphQL request which requires auth.
    /// Returns `None` if the request was not successful.
    session_token: Option<String>,
}

#[derive(juniper::GraphQLObject)]
pub(crate) struct DeauthorizePayload {
    success: bool,
}

/// Webapp is currently using very simplified and minimalistic flow: it basically allows only
/// specific hardcoded Google subjects to be authorized (admins). The thinking is that webapp
/// login is needed only for the backoffice and the access should be limited to admins. This should
/// be easily changeable in the future.
pub(crate) async fn authorize_webapp(
    google_id_token: &str,
    context: &Context,
) -> FieldResult<AuthorizeWebappPayload> {
    let connection_pool = context.pool.to_owned();
    let session_token = crate::auth::authorize(&connection_pool, &google_id_token).await;
    match session_token {
        Ok(session_token) => Ok(AuthorizeWebappPayload {
            success: true,
            failure_message: None,
            session_token: Some(session_token),
        }),
        Err(e) => {
            tracing::error!("{}", e);
            Ok(AuthorizeWebappPayload {
                success: false,
                failure_message: Some(format!("{}", e)),
                session_token: None,
            })
        }
    }
}

pub(crate) async fn deauthorize(
    session_token: &str, // TODO: this could be removed (?) - we can use the user from context
    context: &Context,
) -> DeauthorizePayload {
    let connection_pool = context.pool.to_owned();
    match crate::auth::deauthorize(&connection_pool, &session_token).await {
        Ok(_) => DeauthorizePayload { success: true },
        Err(_) => DeauthorizePayload { success: false },
    }
}
