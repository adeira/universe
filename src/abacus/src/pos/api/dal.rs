use crate::arangodb::{resolve_aql, ConnectionPool};
use crate::commerce::api::SupportedCurrency;
use serde::{Deserialize, Serialize};
use serde_json::json;

#[derive(Serialize, Deserialize, Clone)]
pub struct PosCheckout {
    _id: String,
}

impl PosCheckout {
    pub fn id(&self) -> String {
        self._id.to_owned()
    }
}

#[derive(Serialize, Deserialize)]
pub(in crate::pos) struct PosCheckoutProductInput {
    // Original product ID just in case it's needed one day (even though it might be already
    // deleted). Already deleted product should not be a big deal since we copy here all the
    // necessary product info at the time of the checkout (so current edited product might be
    // completely different and that's OK).
    pub(crate) product_id: String,
    pub(crate) product_name: String,
    pub(crate) product_units: i32,
    pub(crate) product_price_unit_amount: i32,
    pub(crate) product_price_unit_amount_currency: SupportedCurrency,
}

#[derive(Serialize, Deserialize)]
pub(in crate::pos) struct PosCheckoutInput {
    pub(crate) selected_products: Vec<PosCheckoutProductInput>,
}

/// Saves checkout information into database so we know what sales are happening via POS and we can
/// make further decisions based on that. Specifically, we record the following information:
///
/// - date and time of the POS sale
/// - products sold (not only IDs but the whole expanded products so future changes of these
///   products don't affect this POS history)
/// - price for each product at the time of the sale (again, preserving historic state)
pub(in crate::pos) async fn create_checkout(
    pool: &ConnectionPool,
    input: &PosCheckoutInput,
) -> anyhow::Result<PosCheckout> {
    let insert_aql = arangors::AqlQuery::builder()
        .query(
            r#"
            INSERT {
              created_date: DATE_ISO8601(DATE_NOW()),
              selected_products: @selected_products
            } INTO pos_checkouts
            RETURN NEW
            "#,
        )
        .bind_var("selected_products", json!(&input.selected_products))
        .build();

    resolve_aql::<PosCheckout>(&pool, insert_aql).await
}

#[derive(Deserialize, Clone, juniper::GraphQLObject)]
pub(in crate::pos) struct PosCheckoutTotalStats {
    total_checkouts: i32,
    total_sold_units: i32,
    total_sold_unit_amount: i32, // TODO: be careful when calling SUM(â€¦) on different currencies!
}

pub(in crate::pos) async fn get_total_checkout_stats(
    pool: &ConnectionPool,
) -> anyhow::Result<PosCheckoutTotalStats> {
    let stats_aql = arangors::AqlQuery::builder()
        .query(
            r#"
            LET stats_per_checkout = (
              FOR checkout IN pos_checkouts
                LET products = checkout.selected_products
                RETURN {
                  total_checkouts: COUNT(1),
                  total_sold_units: SUM(products[*].product_units),
                  total_sold_unit_amount: SUM(FOR p IN products RETURN (p.product_units * p.product_price_unit_amount)),
                }
            )

            RETURN {
              total_checkouts: SUM(stats_per_checkout[*].total_checkouts),
              total_sold_units: SUM(stats_per_checkout[*].total_sold_units),
              total_sold_unit_amount: SUM(stats_per_checkout[*].total_sold_unit_amount),
            }
            "#,
        )
        .build();

    resolve_aql::<PosCheckoutTotalStats>(&pool, stats_aql).await
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::arangodb::{cleanup_test_database, prepare_empty_test_database};

    #[ignore]
    #[tokio::test]
    async fn create_checkout_test() {
        let db_name = "create_checkout_test";
        let pool = prepare_empty_test_database(&db_name).await;

        // 1) create first checkout
        create_checkout(
            &pool,
            &PosCheckoutInput {
                selected_products: vec![
                    PosCheckoutProductInput {
                        product_id: String::from("products/1"),
                        product_name: String::from("Product name 1"),
                        product_units: 1,
                        product_price_unit_amount: 120,
                        product_price_unit_amount_currency: SupportedCurrency::MXN,
                    },
                    PosCheckoutProductInput {
                        product_id: String::from("products/2"),
                        product_name: String::from("Product name 2"),
                        product_units: 2,
                        product_price_unit_amount: 150,
                        product_price_unit_amount_currency: SupportedCurrency::MXN,
                    },
                ],
            },
        )
        .await
        .unwrap();

        // 2) create second checkout
        create_checkout(
            &pool,
            &PosCheckoutInput {
                selected_products: vec![PosCheckoutProductInput {
                    product_id: String::from("products/3"),
                    product_name: String::from("Product name 3"),
                    product_units: 5,
                    product_price_unit_amount: 200,
                    product_price_unit_amount_currency: SupportedCurrency::MXN,
                }],
            },
        )
        .await
        .unwrap();

        // 3) return the checkout stats
        let checkout_stats = get_total_checkout_stats(&pool).await.unwrap();
        assert_eq!(checkout_stats.total_checkouts, 2); // first and second
        assert_eq!(checkout_stats.total_sold_units, 8); // 1 + 2 + 5
        assert_eq!(checkout_stats.total_sold_unit_amount, 1420); // (120 + 2*150 + 5*200)

        cleanup_test_database(&db_name).await;
    }
}
