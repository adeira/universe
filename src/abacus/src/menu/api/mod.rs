use crate::commerce::api::{Product, SupportedLocale};
use crate::graphql::AbacusGraphQLResult;
use crate::graphql_context::Context;
use crate::menu::MenuSections;

pub(crate) struct MenuQuery;

#[juniper::graphql_object(context = Context)]
impl MenuQuery {
    /// Returns a specified section of our menu. This is to maintain one source of truth (in our
    /// database) about the prices, descriptions, translations and similar.
    async fn menu(
        context: &Context,
        client_locale: SupportedLocale,
        section: MenuSections,
    ) -> AbacusGraphQLResult<Vec<Product>> {
        Ok(crate::menu::get_section(&context, &client_locale, &section).await?)
    }
}
