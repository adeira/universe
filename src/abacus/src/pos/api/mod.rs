use crate::auth::rbac;
use crate::auth::rbac::Actions::Pos;
use crate::auth::rbac::PosActions::{Checkout, GetAllPublishedProducts, GetCheckoutStats};
use crate::commerce::api::{
    PriceSortDirection, Product, ProductMultilingualInputVisibility, SupportedCurrency,
    SupportedLocale,
};
use crate::graphql_context::Context;
use crate::pos::api::dal::{
    create_checkout, get_total_checkout_stats, PosCheckoutInput as PosCheckoutDalInput,
    PosCheckoutProductInput as PosCheckoutProductDalInput, PosCheckoutTotalStats,
};

pub(crate) struct POSQuery;
pub(crate) struct POSMutation;

mod dal;

#[juniper::graphql_object(context = Context)]
impl POSQuery {
    /// Lists published products for POS. Requires admin permissions so it should be used only in
    /// POS after logging in.
    async fn list_published_products(context: &Context) -> Option<Vec<Option<Product>>> {
        match rbac::verify_permissions(&context.user, &Pos(GetAllPublishedProducts)).await {
            Ok(_) => {
                // only admin can list the products (POS is private)
                crate::commerce::api::search_published_products(
                    &context,
                    &SupportedLocale::EnUS,         // TODO
                    &PriceSortDirection::HighToLow, // TODO
                    &None,
                    &ProductMultilingualInputVisibility::POS,
                )
                .await
            }
            Err(_) => None,
        }
    }

    async fn get_total_checkout_stats(context: &Context) -> Option<PosCheckoutTotalStats> {
        match rbac::verify_permissions(&context.user, &Pos(GetCheckoutStats)).await {
            Ok(_) => {
                // only admin can read the checkout stats
                // TODO: do not expose this directly but hide it into a model layer
                match get_total_checkout_stats(&context.pool).await {
                    Ok(total_checkout_stats) => Some(total_checkout_stats),
                    Err(_) => None, // TODO
                }
            }
            Err(_) => None, // TODO
        }
    }
}

#[derive(juniper::GraphQLObject)]
pub struct PosCheckoutPayload {
    id: juniper::ID,
}

#[derive(juniper::GraphQLObject)]
pub struct PosCheckoutError {
    message: String,
}

#[derive(juniper::GraphQLUnion)]
pub enum PosCheckoutPayloadOrError {
    Payload(PosCheckoutPayload),
    Error(PosCheckoutError),
}

#[derive(juniper::GraphQLInputObject)]
pub struct PosCheckoutProductInput {
    pub(crate) product_key: juniper::ID,
    pub(crate) product_units: i32,
    pub(crate) product_price_unit_amount: i32,
    pub(crate) product_price_unit_amount_currency: SupportedCurrency,
}

#[derive(juniper::GraphQLInputObject)]
pub struct PosCheckoutInput {
    pub(crate) selected_products: Vec<PosCheckoutProductInput>,
}

#[juniper::graphql_object(context = Context)]
impl POSMutation {
    /// This is a simplified POS checkout. We simply record what the user bought for how much and
    /// so on. There is almost no validation - what the client sends is what we record. This is the
    /// main difference from eshop checkout where we would have to verify the prices for example.
    ///
    /// Why not to verify that the checkout price matches the product price? It's because when
    /// cashier accepts the money, the product is sold for the given price and there is not time for
    /// price adjustments (customers would be angry if we would say "oh, actually it just got more
    /// expensive").
    async fn checkout(context: &Context, input: PosCheckoutInput) -> PosCheckoutPayloadOrError {
        // Contrary to what DAL requires, we accept only product keys in GraphQL and expand them on BE.

        let product_keys = input
            .selected_products
            .iter()
            .map(|product| product.product_key.to_string())
            .collect::<Vec<String>>();

        let products = crate::commerce::api::get_products_by_keys(
            &context,
            &SupportedLocale::EnUS, // TODO
            &product_keys,
        )
        .await;

        let selected_products = match products {
            Ok(products) => products
                .iter()
                .map(|product| {
                    let checkout_product = input
                        .selected_products
                        .iter()
                        .find(|p| p.product_key.to_string() == product.key())
                        .unwrap();

                    // Most of the values should be taken from the POS input except some product
                    // details which are being expanded from the product ID.
                    PosCheckoutProductDalInput {
                        product_id: product.id(),
                        product_name: product.name(),
                        product_units: checkout_product.product_units,
                        product_price_unit_amount: checkout_product.product_price_unit_amount,
                        product_price_unit_amount_currency: checkout_product
                            .product_price_unit_amount_currency,
                    }
                })
                .collect(),
            Err(e) => {
                return PosCheckoutPayloadOrError::Error(PosCheckoutError {
                    message: format!("{:?}", e),
                });
            }
        };

        match rbac::verify_permissions(&context.user, &Pos(Checkout)).await {
            Ok(_) => {
                match create_checkout(&context.pool, &PosCheckoutDalInput { selected_products })
                    .await
                {
                    Ok(checkout) => PosCheckoutPayloadOrError::Payload(PosCheckoutPayload {
                        id: juniper::ID::from(checkout.id()),
                    }),
                    Err(e) => PosCheckoutPayloadOrError::Error(PosCheckoutError {
                        message: format!("{:?}", e),
                    }),
                }
            }
            Err(_) => PosCheckoutPayloadOrError::Error(PosCheckoutError {
                message: String::from("not enough permissions to perform POS checkout"),
            }),
        }
    }
}
