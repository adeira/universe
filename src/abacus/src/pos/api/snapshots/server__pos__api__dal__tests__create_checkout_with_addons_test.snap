---
source: src/pos/api/dal.rs
expression: all_checkouts
---
[
  PosCheckout(
    _id: "[REDACTED]",
    created_date: "[REDACTED]",
    selected_products: [
      PosCheckoutProductInput(
        product_id: "products/1",
        product_name: "Product name 1",
        product_units: 1,
        product_price_unit_amount: 100,
        product_price_unit_amount_currency: MXN,
        product_addons: None,
      ),
      PosCheckoutProductInput(
        product_id: "products/1",
        product_name: "Product name 1 - with addons",
        product_units: 1,
        product_price_unit_amount: 100,
        product_price_unit_amount_currency: MXN,
        product_addons: Some([
          PosCheckoutProductAddonInput(
            product_addon_id: "product_addons/1",
            product_addon_extra_price_unit_amount: 10,
            product_addon_extra_price_unit_amount_currency: MXN,
          ),
          PosCheckoutProductAddonInput(
            product_addon_id: "product_addons/2",
            product_addon_extra_price_unit_amount: 20,
            product_addon_extra_price_unit_amount_currency: MXN,
          ),
        ]),
      ),
    ],
  ),
]
